File: C++/8-Graphs(BFS-DFS)/Algorithms/4-Floyd-WarshallAlgorithm/Floyd-WarshallAlgorithmPro.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <iomanip>
#include <algorithm>
#include <queue>
#include <limits> // Required for numeric_limits

using namespace std;

// Function to print the distance matrix
void printSolution(const vector<vector<int>>& dist, const string& title) {
    cout << "\n" << title << ":\n";
    for (const auto& row : dist) {
        for (int val : row) {
            if (val == INT_MAX)
                cout << setw(5) << "INF"; // Use setw for consistent spacing
            else
                cout << setw(5) << val;
        }
        cout << endl;
    }
}

// Function to initialize the distance matrix
void initializeDistanceMatrix(vector<vector<int>>& dist, const vector<vector<int>>& graph) {
    int V = graph.size();
    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            if (i == j) {
                dist[i][j] = 0; // Distance from a node to itself is 0
            } else if (graph[i][j] != 0) {
                dist[i][j] = graph[i][j]; // Use the edge weight from the graph
            } else {
                dist[i][j] = INT_MAX; // No direct edge, initialize to infinity
            }
        }
    }
}

// 1. Basic Floyd-Warshall Algorithm
//   - Computes all-pairs shortest paths in a graph.
//   - Useful for route planning in GPS systems.
void floydWarshallBasic(vector<vector<int>>& dist, int V) {
    for (int k = 0; k < V; ++k) {
        for (int i = 0; i < V; ++i) {
            for (int j = 0; j < V; ++j) {
                if (dist[i][k] != INT_MAX && dist[k][j] != INT_MAX) { // Check for overflow
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
    }
    //check for negative weight cycles.  If any diagonal element is negative, there is a negative weight cycle.
    for (int i = 0; i < V; i++) {
        if (dist[i][i] < 0) {
            cout << "Negative weight cycle detected!" << endl;
            return;
        }
    }
}

// 2. Floyd-Warshall with Path Reconstruction
//   - Extends the basic algorithm to also compute the shortest paths.
//   - Useful for network routing protocols.
void floydWarshallWithPathReconstruction(vector<vector<int>>& dist, vector<vector<int>>& next, int V, const vector<vector<int>>& graph) {
     // Initialize the 'next' matrix.  Stores the intermediate node in the shortest path
    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            if (i == j)
                next[i][j] = i;  // Path from i to i is just i
            else if (graph[i][j] != 0)
                next[i][j] = j; // Initial next node is j
            else
                next[i][j] = -1; // No direct path
        }
    }

    for (int k = 0; k < V; ++k) {
        for (int i = 0; i < V; ++i) {
            for (int j = 0; j < V; ++j) {
                if (dist[i][k] != INT_MAX && dist[k][j] != INT_MAX && dist[i][k] + dist[k][j] < dist[i][j]) {
                    dist[i][j] = dist[i][k] + dist[k][j];
                    next[i][j] = next[i][k]; // Update the intermediate node
                }
            }
        }
    }
    //check for negative weight cycles.
    for (int i = 0; i < V; i++) {
        if (dist[i][i] < 0) {
            cout << "Negative weight cycle detected!" << endl;
            return;
        }
    }
}

// Function to print the shortest path
void printPath(const vector<vector<int>>& next, int u, int v) {
    if (next[u][v] == -1) {
        cout << "No path exists from " << u << " to " << v << endl;
        return;
    }
    cout << "Shortest path from " << u << " to " << v << ": " << u << " ";
    while (u != v) {
        u = next[u][v];
        cout << u << " ";
    }
    cout << endl;
}

// 3. Floyd-Warshall for Transitive Closure
//   - Determines if there is a path between any two vertices in a graph.
//   - Useful for analyzing relationships in social networks.
void floydWarshallTransitiveClosure(vector<vector<int>>& reachability, int V, const vector<vector<int>>& graph) {
    // Initialize reachability matrix.  0/1 instead of distances.
    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            if (i == j || graph[i][j] != 0)
                reachability[i][j] = 1; // There is a path from a node to itself, and to direct neighbors
            else
                reachability[i][j] = 0;
        }
    }

    for (int k = 0; k < V; ++k) {
        for (int i = 0; i < V; ++i) {
            for (int j = 0; j < V; ++j) {
                reachability[i][j] = reachability[i][j] || (reachability[i][k] && reachability[k][j]);
            }
        }
    }
}

// 4. Floyd-Warshall for Detecting Negative Cycles
//    - Detects the presence of negative weight cycles in a graph.
//    - Important for currency exchange arbitrage detection.
void floydWarshallNegativeCycleDetection(vector<vector<int>>& dist, int V, bool& hasNegativeCycle) {
     for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            if (i == j) {
                dist[i][j] = 0; // Distance from a node to itself is 0
            }
            else if (dist[i][j] == 0)
                dist[i][j] = INT_MAX;
        }
    }
    for (int k = 0; k < V; ++k) {
        for (int i = 0; i < V; ++i) {
            for (int j = 0; j < V; ++j) {
                if (dist[i][k] != INT_MAX && dist[k][j] != INT_MAX)
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
            }
        }
    }

    hasNegativeCycle = false;
    for (int i = 0; i < V; ++i) {
        if (dist[i][i] < 0) {
            hasNegativeCycle = true;
            break;
        }
    }
}

// 5. Floyd-Warshall with Arbitrary Precision (using long double)
//    - Handles larger edge weights or more precision to avoid overflow.
//    - Useful in financial modeling or scientific simulations.
void floydWarshallArbitraryPrecision(vector<vector<long double>>& dist_ld, int V, const vector<vector<int>>& graph) {
    // Initialize the distance matrix with long doubles.
    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            if (i == j) {
                dist_ld[i][j] = 0.0;
            } else if (graph[i][j] != 0) {
                dist_ld[i][j] = static_cast<long double>(graph[i][j]);
            } else {
                dist_ld[i][j] = numeric_limits<long double>::infinity();
            }
        }
    }

    for (int k = 0; k < V; ++k) {
        for (int i = 0; i < V; ++i) {
            for (int j = 0; j < V; ++j) {
                if (dist_ld[i][k] != numeric_limits<long double>::infinity() && dist_ld[k][j] != numeric_limits<long double>::infinity()) {
                    dist_ld[i][j] = min(dist_ld[i][j], dist_ld[i][k] + dist_ld[k][j]);
                }
            }
        }
    }
    for (int i = 0; i < V; i++) {
        if (dist_ld[i][i] < 0) {
            cout << "Negative weight cycle detected!" << endl;
            return;
        }
    }
}

int main() {
    // Example graph represented as an adjacency matrix
    vector<vector<int>> graph = {
        {0,   5,  INT_MAX, 10},
        {INT_MAX, 0,   3,  INT_MAX},
        {INT_MAX, INT_MAX, 0,   1},
        {INT_MAX, INT_MAX, INT_MAX, 0}
    };
    int V = graph.size();

    // 1. Basic Floyd-Warshall
    vector<vector<int>> dist_basic(V, vector<int>(V));
    initializeDistanceMatrix(dist_basic, graph);
    floydWarshallBasic(dist_basic, V);
    printSolution(dist_basic, "1. Basic Floyd-Warshall");

    // 2. Floyd-Warshall with Path Reconstruction
    vector<vector<int>> dist_path(V, vector<int>(V));
    vector<vector<int>> next(V, vector<int>(V));
    initializeDistanceMatrix(dist_path, graph);
    floydWarshallWithPathReconstruction(dist_path, next, V, graph);
    printSolution(dist_path, "2. Floyd-Warshall with Path Reconstruction");
    printPath(next, 0, 3); // Example path

    // 3. Floyd-Warshall for Transitive Closure
    vector<vector<int>> reachability(V, vector<int>(V));
     floydWarshallTransitiveClosure(reachability, V, graph);
    printSolution(reachability, "3. Floyd-Warshall for Transitive Closure");

    // 4. Floyd-Warshall for Detecting Negative Cycles
    vector<vector<int>> dist_negative_cycle(V, vector<int>(V));
    bool hasNegativeCycle;
    initializeDistanceMatrix(dist_negative_cycle, graph); // Important to initialize *before* calling the function
    floydWarshallNegativeCycleDetection(dist_negative_cycle, V, hasNegativeCycle);
    if (hasNegativeCycle)
        cout << "4. Graph contains a negative weight cycle" << endl;
    else
        cout << "4. Graph does not contain any negative weight cycle" << endl;

     // 5. Floyd-Warshall with Arbitrary Precision
    vector<vector<long double>> dist_ld(V, vector<long double>(V));
    floydWarshallArbitraryPrecision(dist_ld, V, graph);
    cout << "\n5. Floyd-Warshall with Arbitrary Precision:\n";
    for (const auto& row : dist_ld) {
        for (long double val : row) {
             if (val == numeric_limits<long double>::infinity())
                cout << setw(10) << "INF";
            else
                cout << setw(10) << fixed << setprecision(2) << val;
        }
        cout << endl;
    }

    return 0;
}

File: Java/8-Graphs(BFS-DFS)/Algorithms/4-Floyd-WarshallAlgorithm/Floyd-WarshallAlgorithmPro.java
import java.util.Arrays;

public class FloydWarshall {

    // Define a constant for infinity, used to represent unreachable nodes.  Using Integer.MAX_VALUE directly can lead to overflow issues
    // when adding to it, so we use a slightly smaller value.
    private static final int INF = 1000000000; // A large number, but not too close to Integer.MAX_VALUE

    public static void main(String[] args) {
        // Example graph represented as an adjacency matrix.  The value at matrix[i][j] represents the weight of the edge
        // from node i to node j.  If there is no direct edge, we use INF.
        int[][] graph = {
                {0, 5, INF, 10},
                {INF, 0, 3, INF},
                {INF, INF, 0, 1},
                {INF, INF, INF, 0}
        };

        System.out.println("Original Graph:");
        printMatrix(graph);

        // 1. Basic Floyd-Warshall Algorithm:  Finds the shortest paths between all pairs of nodes in the graph.
        int[][] shortestPaths1 = floydWarshallBasic(graph);
        System.out.println("\n1. Basic Floyd-Warshall Algorithm:");
        printMatrix(shortestPaths1);

        // 2. Floyd-Warshall with Path Reconstruction:  Computes the shortest paths and also determines the intermediate
        // nodes in those paths, allowing us to reconstruct the full path.
        int[][] shortestPaths2 = floydWarshallWithPathReconstruction(graph);
        System.out.println("\n2. Floyd-Warshall with Path Reconstruction:");
        printMatrix(shortestPaths2);
        // Example of how to reconstruct a path from node 0 to node 3
        System.out.println("Example Path from 0 to 3: " + reconstructPath(graph, 0, 3));


        // 3. Floyd-Warshall for Detecting Negative Cycles:  Checks if the graph contains any negative-weight cycles.  If a negative
        // cycle exists, the shortest path between some nodes can become infinitely small (negative), which can cause problems
        // in some applications.
        boolean hasNegativeCycle = floydWarshallNegativeCycleDetection(graph);
        System.out.println("\n3. Floyd-Warshall for Detecting Negative Cycles:");
        System.out.println("Has Negative Cycle: " + hasNegativeCycle);

        // 4. Floyd-Warshall with Optimization (Space):  The basic Floyd-Warshall algorithm uses a 2D array to store the
        // intermediate distances.  We can optimize this to use only a 1D array, reducing the space complexity.
        int[] shortestPaths4 = floydWarshallOptimizedSpace(graph);
        System.out.println("\n4. Floyd-Warshall with Optimization (Space):");
        printArrayAsMatrix(shortestPaths4, graph.length); // Print 1D array as matrix

        // 5. Floyd-Warshall for finding the Transitive Closure: Finds if there is a path between all pairs of vertices
        // A transitive closure matrix has 1 at i,j if a path exists from i to j, otherwise 0.
        int[][] transitiveClosure = transitiveClosure(graph);
        System.out.println("\n5. Floyd-Warshall for Transitive Closure:");
        printMatrix(transitiveClosure);
    }

    /**
     * 1. Basic Floyd-Warshall Algorithm Implementation:
     *
     * Computes the shortest distances between all pairs of nodes in a graph.
     *
     * @param graph A square matrix representing the graph, where graph[i][j] is the weight of the edge from node i to node j.
     * Use INF (a large value) to represent no direct edge.  The graph matrix is assumed to represent
     * edge weights, and the algorithm will modify it to store shortest path distances.
     * @return A new matrix representing the shortest distances between all pairs of nodes.  The original graph
     * matrix is NOT modified by this method.
     */
    public static int[][] floydWarshallBasic(int[][] graph) {
        int V = graph.length; // Number of vertices in the graph.
        // Create a copy of the graph to store the shortest path distances.  This is important to avoid modifying
        // the original graph data.
        int[][] dist = new int[V][V];
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                dist[i][j] = graph[i][j];
            }
        }

        // The core of the Floyd-Warshall algorithm.  It iterates through all possible intermediate nodes (k).
        // For each pair of nodes (i, j), it checks if going through k is shorter than the current shortest path
        // from i to j.
        for (int k = 0; k < V; k++) {
            for (int i = 0; i < V; i++) {
                for (int j = 0; j < V; j++) {
                    // Check for overflow before adding.  If either dist[i][k] or dist[k][j] is INF, adding them could
                    // result in a value that wraps around to a small positive number, leading to incorrect results.
                    if (dist[i][k] != INF && dist[k][j] != INF) {
                        dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                    }
                }
            }
        }
        return dist;
    }

    /**
     * 2. Floyd-Warshall Algorithm with Path Reconstruction:
     *
     * Computes the shortest distances between all pairs of nodes and also stores the intermediate nodes
     * used in the shortest paths.  This allows for reconstructing the actual path taken.
     *
     * @param graph A square matrix representing the graph (same format as floydWarshallBasic).
     * @return A new matrix representing the shortest distances between all pairs of nodes.
     */

    public static int[][] floydWarshallWithPathReconstruction(int[][] graph) {
        int V = graph.length;
        int[][] dist = new int[V][V];
        int[][] next = new int[V][V]; // Stores the intermediate node in the shortest path from i to j.

        // Initialize the distance matrix and the 'next' matrix.
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                dist[i][j] = graph[i][j];
                if (graph[i][j] != INF && i != j) {
                    next[i][j] = j; // Initially, the next node is j itself (direct edge).
                } else {
                    next[i][j] = -1; // No intermediate node.
                }
            }
        }

        // Core of the algorithm, same as before, but also update the 'next' matrix.
        for (int k = 0; k < V; k++) {
            for (int i = 0; i < V; i++) {
                for (int j = 0; j < V; j++) {
                    if (dist[i][k] != INF && dist[k][j] != INF && dist[i][k] + dist[k][j] < dist[i][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                        next[i][j] = next[i][k]; // The next node on the path from i to j is now the same as the next node on the path from i to k.
                    }
                }
            }
        }
        return dist;
    }

    /**
     * Helper function to reconstruct the shortest path between two nodes, using the 'next' matrix
     * computed by floydWarshallWithPathReconstruction.
     *
     * @param graph The original graph (used to check for valid nodes).
     * @param start The starting node.
     * @param end   The ending node.
     * @return A list of the nodes in the shortest path from start to end, or an empty list if no path exists.
     */
    public static String reconstructPath(int[][] graph, int start, int end) {
        int V = graph.length;
        if (start < 0 || start >= V || end < 0 || end >= V) {
            return "Invalid input: Start or end node out of bounds.";
        }

        int[][] next = new int[V][V]; // Create the 'next' matrix.
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                if (graph[i][j] != INF && i != j) {
                    next[i][j] = j;
                } else {
                    next[i][j] = -1;
                }
            }
        }

        for (int k = 0; k < V; k++) {
            for (int i = 0; i < V; i++) {
                for (int j = 0; j < V; j++) {
                    if (graph[i][k] != INF && graph[k][j] != INF && graph[i][k] + graph[k][j] < graph[i][j]) {
                        next[i][j] = next[i][k];
                    }
                }
            }
        }
        if (next[start][end] == -1) {
            return "No path exists between " + start + " and " + end;
        } else {
            StringBuilder path = new StringBuilder();
            path.append(start);
            int current = start;
            while (current != end) {
                current = next[current][end];
                path.append(" -> ").append(current);
            }
            return path.toString();
        }
    }


    /**
     * 3. Floyd-Warshall Algorithm for Detecting Negative Cycles:
     *
     * Detects if a graph contains any negative-weight cycles.  A negative-weight cycle is a cycle in the graph
     * where the sum of the edge weights is negative.  The presence of a negative cycle means that there is no
     * shortest path between some nodes, as you can keep going around the cycle to decrease the path length
     * indefinitely.
     *
     * @param graph A square matrix representing the graph.
     * @return true if the graph contains a negative cycle, false otherwise.
     */
    public static boolean floydWarshallNegativeCycleDetection(int[][] graph) {
        int V = graph.length;
        int[][] dist = new int[V][V];
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                dist[i][j] = graph[i][j];
            }
        }

        for (int k = 0; k < V; k++) {
            for (int i = 0; i < V; i++) {
                for (int j = 0; j < V; j++) {
                    if (dist[i][k] != INF && dist[k][j] != INF) {
                        dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                    }
                }
            }
        }

        // After the main loop, check the diagonal elements of the distance matrix.  If any diagonal element is negative,
        // it means there is a negative-weight cycle reachable from that node.  This is because the diagonal element
        // represents the shortest path from a node to itself.  In a graph with no negative cycles, this should always be 0.
        for (int i = 0; i < V; i++) {
            if (dist[i][i] < 0) {
                return true; // Negative cycle detected.
            }
        }
        return false; // No negative cycle found.
    }

    /**
     * 4. Floyd-Warshall Algorithm with Optimization (Space):
     *
     * Optimizes the space complexity of the Floyd-Warshall algorithm from O(V^2) to O(V) by using a single 1D array
     * instead of a 2D array.  This is possible because the algorithm iteratively updates the distances, and we
     * can overwrite the previous distances in place.
     *
     * @param graph A square matrix representing the graph.
     * @return A 1D array representing the shortest distances between all pairs of nodes.  The array is organized such
     * that the distances from node i to all other nodes are stored in consecutive elements.  Specifically,
     * the distance from node i to node j is stored at index i * V + j, where V is the number of vertices.
     */
    public static int[] floydWarshallOptimizedSpace(int[][] graph) {
        int V = graph.length;
        int[] dist = new int[V * V]; // 1D array to store distances.

        // Initialize the 1D array from the 2D graph matrix.
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                dist[i * V + j] = graph[i][j];
            }
        }

        // Core of the algorithm, adapted for the 1D array representation.
        for (int k = 0; k < V; k++) {
            for (int i = 0; i < V; i++) {
                for (int j = 0; j < V; j++) {
                    if (dist[i * V + k] != INF && dist[k * V + j] != INF) {
                        dist[i * V + j] = Math.min(dist[i * V + j], dist[i * V + k] + dist[k * V + j]);
                    }
                }
            }
        }
        return dist;
    }

    /**
     * 5. Floyd-Warshall Algorithm for Transitive Closure:
     *
     * Computes the transitive closure of a graph.  The transitive closure is a matrix that indicates whether
     * there is a path (of any length) between any two nodes in the graph.  If there is a path from node i to node j,
     * the (i, j) entry in the transitive closure matrix is 1; otherwise, it is 0.
     *
     * @param graph A square matrix representing the graph.
     * @return A new matrix representing the transitive closure of the graph.
     */
    public static int[][] transitiveClosure(int[][] graph) {
        int V = graph.length;
        int[][] tc = new int[V][V];

        // Initialize the transitive closure matrix.  If there is an edge from i to j, then there is a path from i to j.
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                if (graph[i][j] != INF || i == j) { // Important: i == j means there is a path from a node to itself.
                    tc[i][j] = 1;
                } else {
                    tc[i][j] = 0;
                }
            }
        }

        // Floyd-Warshall algorithm adapted for transitive closure.  Instead of computing shortest distances, we're
        // checking for path existence.  If there is a path from i to k AND a path from k to j, then there is a path from i to j.
        for (int k = 0; k < V; k++) {
            for (int i = 0; i < V; i++) {
                for (int j = 0; j < V; j++) {
                    tc[i][j] = tc[i][j] | (tc[i][k] & tc[k][j]); // If there is a path i->k AND k->j, then there is a path i->j
                }
            }
        }
        return tc;
    }

    /**
     * Helper function to print a 2D matrix.
     *
     * @param matrix The matrix to print.
     */
    public static void printMatrix(int[][] matrix) {
        for (int[] row : matrix) {
            for (int val : row) {
                if (val == INF) {
                    System.out.print("INF\t");
                } else {
                    System.out.print(val + "\t");
                }
            }
            System.out.println();
        }
    }

    /**
     * Helper function to print a 1D array as a 2D matrix.  Useful for displaying the output of
     * floydWarshallOptimizedSpace.
     *
     * @param array The 1D array to print.
     * @param V     The number of vertices (used to determine the dimensions of the matrix).
     */
    public static void printArrayAsMatrix(int[] array, int V) {
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                int val = array[i * V + j];
                if (val == INF) {
                    System.out.print("INF\t");
                } else {
                    System.out.print(val + "\t");
                }
            }
            System.out.println();
        }
    }
}

